nocturne/envs/vec_env_ma.py

```diff
@@ -0,0 +1,242 @@
+"""Vectorized environment wrapper for multi-agent environments."""
+import logging
+import time
+from collections import OrderedDict
+from copy import deepcopy
+from typing import Any, Dict, List
+
+import gymnasium as gym
+import numpy as np
+from stable_baselines3.common.vec_env.base_vec_env import (
+    VecEnv,
+    VecEnvObs,
+    VecEnvStepReturn,
+)
+from stable_baselines3.common.vec_env.util import (
+    copy_obs_dict,
+    dict_to_obs,
+    obs_space_info,
+)
+
+from nocturne.envs.base_env import BaseEnv
+from utils.config import load_config
+
+logging.basicConfig(level=logging.INFO)
+
+
+class MultiAgentAsVecEnv(VecEnv):
+    """A wrapper that casts multi-agent environments as vectorized environments.
+
+    Args:
+    -----
+        VecEnv (SB3 VecEnv): SB3 VecEnv base class.
+    """
+
+    def __init__(self, config, num_envs):
+        # Create Nocturne env
+        self.env = BaseEnv(config)
+
+        # Make action and observation spaces compatible with SB3 (requires gymnasium)
+        self.action_space = gym.spaces.Discrete(self.env.action_space.n)
+        self.observation_space = gym.spaces.Box(-np.inf, np.inf, self.env.observation_space.shape, np.float32)
+        self.num_envs = num_envs  # The maximum number of agents allowed in the environment
+        self.keys, shapes, dtypes = obs_space_info(self.env.observation_space)
+
+        # Storage
+        self.buf_obs = OrderedDict(
+            [(k, np.zeros((self.num_envs, *tuple(shapes[k])), dtype=dtypes[k])) for k in self.keys]
+        )
+        self.buf_dones = np.full(fill_value=np.nan, shape=(self.num_envs,))
+        self.buf_rews = np.full(fill_value=np.nan, shape=(self.num_envs,))
+        self.buf_infos: List[Dict[str, Any]] = [{} for _ in range(self.num_envs)]
+        self.n_episodes = 0
+        self.episode_lengths = []
+        self.rewards = []  # Log reward per step
+        self.dead_agent_ids = []  # Log dead agents per step
+        self.frac_collided = []  # Log fraction of agents that collided
+        self.frac_goal_achieved = []  # Log fraction of agents that achieved their goal
+        self.agents_in_scene = []
```

Comment:
This is not important now but this should probably be separated out into a separate logger object